editoast/src/core/v2/pathfinding.rs

```diff
@@ -93,13 +98,84 @@ pub enum PathfindingResult {
     },
 }
 
-#[derive(Serialize, Deserialize, Clone, Debug, ToSchema, Hash)]
+/// An oriented range on a track section.
+/// `begin` is always less than `end`.
+#[derive(Serialize, Deserialize, Clone, Debug, ToSchema, Hash, PartialEq, Eq)]
 pub struct TrackRange {
+    /// The track section identifier.
     #[schema(inline)]
-    track_section: Identifier,
-    begin: u64,
-    end: u64,
-    direction: Direction,
+    pub track_section: Identifier,
+    /// The beginning of the range in mm.
+    pub begin: u64,
+    /// The end of the range in mm.
+    pub end: u64,
+    /// The direction of the range.
+    pub direction: Direction,
+}
+
+impl TrackRange {
+    #[cfg(test)]
+    /// Creates a new `TrackRange`.
+    pub fn new<T: AsRef<str>>(
+        track_section: T,
+        begin: u64,
+        end: u64,
+        direction: Direction,
+    ) -> Self {
+        Self {
+            track_section: track_section.as_ref().into(),
+            begin,
+            end,
+            direction,
+        }
+    }
+
+    /// Returns the starting offset of the range (depending on the direction).
+    pub fn start(&self) -> u64 {
+        if self.direction == Direction::StartToStop {
+            self.begin
+        } else {
+            self.end
+        }
+    }
+
+    /// Returns the ending offset of the range (depending on the direction).
+    pub fn stop(&self) -> u64 {
+        if self.direction == Direction::StartToStop {
+            self.end
+        } else {
+            self.begin
+        }
+    }
+
+    /// Computes a TrackRangeOffset location on this track range following its direction
+    pub fn offset(&self, offset: u64) -> TrackRangeOffset<'_> {
+        if self.direction == Direction::StartToStop {
+            return TrackRangeOffset {
+                track_range: self,
+                offset: self.begin + offset,
+            };
+        }
+        TrackRangeOffset {
+            track_range: self,
+            offset: self.end - offset,
+        }
```

Comment:
```suggestion
        assert!(offset <= self.length(), "offset out of track range bounds");
        TrackRangeOffset {
            track_range: self,
            offset,
        }
```

In the suggestion I made yesterday in `get_location` I supposed that the offset was relative to the track range (i.e. `track_range_offset == 0 => track_offset = track_range.begin`. So I'm suspecting that this part of `get_location` doesn't work as intended anymore.